compiler/rustc_smir/src/rustc_internal/pretty.rs

```diff
@@ -0,0 +1,308 @@
+use std::io;
+
+use rustc_middle::ty::TyCtxt;
+use stable_mir::{
+    mir::{Mutability, Operand, Rvalue, StatementKind},
+    ty::{RigidTy, TyKind},
+    CrateItem,
+};
+
+use super::{internal, run};
+
+pub fn write_smir_pretty<'tcx>(tcx: TyCtxt<'tcx>, w: &mut dyn io::Write) -> io::Result<()> {
+    writeln!(w, "// WARNING: This is highly experimental output it's intended for stable-mir developers only.").unwrap();
+    writeln!(w, "// If you find a bug or want to improve the output open a issue at https://github.com/rust-lang/project-stable-mir.").unwrap();
+
+    run(tcx, || {
+        let items = stable_mir::all_local_items();
+        let _ = items
+            .iter()
+            .map(|item| -> io::Result<()> {
+                // Because we can't return a Result from a closure, we have to unwrap here.
+                writeln!(w, "{}", function_name(*item, tcx))?;
+                writeln!(w, "{}", function_body(*item, tcx))?;
+                let _ = item
+                    .body()
+                    .blocks
+                    .iter()
+                    .enumerate()
+                    .map(|(index, block)| -> io::Result<()> {
+                        writeln!(w, "    bb{}: {{", index)?;
+                        let _ = block
+                            .statements
+                            .iter()
+                            .map(|statement| -> io::Result<()> {
+                                writeln!(w, "{}", pretty_statement(&statement.kind, tcx))?;
+                                Ok(())
+                            })
+                            .collect::<Vec<_>>();
+                        writeln!(w, "    }}").unwrap();
+                        Ok(())
+                    })
+                    .collect::<Vec<_>>();
+                Ok(())
+            })
+            .collect::<Vec<_>>();
+    });
+    Ok(())
+}
```

Comment:
What about something like this (btw, don't accept it as is, as it probably has some typos):

```suggestion
pub fn write_smir_pretty<'tcx, W: io::Write>(tcx: TyCtxt<'tcx>, w: &mut W) -> io::Result<()> {
    writeln!(w, "// WARNING: This is highly experimental output it's intended for stable-mir developers only.")?;
    writeln!(w, "// If you find a bug or want to improve the output open a issue at https://github.com/rust-lang/project-stable-mir.")?;

    run(tcx, || {
        let items = stable_mir::all_local_items();
        items
            .iter()
            .map(|item| -> io::Result<()> {
                item.dump()
            })
            .collect::<Result<_>>()
    })?;
    Ok(())
}

// This should go in the stable_mir crate.
impl CrateItem {
    pub fn dump<W: io::Write>(&self, w: &mut W)  -> io::Result<()> {
                writeln!(w, "{}", function_header(*item))?;
                body().dump(w)
    }
}

impl Body {
    pub fn dump<W: io::Write>(&self, w: &mut W)  -> io::Result<()> {
                // TODO: print_locals as in function_body()
                
                self
                    .blocks
                    .iter()
                    .enumerate()
                    .map(|(index, block)| -> io::Result<()> {
                        writeln!(w, "    bb{}: {{", index)?;
                        block
                            .statements
                            .iter()
                            .map(|statement| -> io::Result<()> {
                                writeln!(w, "{}", pretty_statement(&statement.kind))
                            })
                            .collect::<Result<_>>()?;
                        writeln!(w, "    }}")
                    })
                    .collect::<Result<_>>()?;
                Ok(())
    }
}
```

Reply:
I like this idea, imo this will work but I just need to hijack your `def_ty()` function before your PR merges, because I can't call any `tcx` functions inside stable mir 🤣

Reply:
Well that required more than copying this function, I'm going to leave `Adt` as a debug right now and we can fix it later once your pr merges.